### PR TITLE
DT-235 move to cell

### PR DESCRIPTION
**Pull request by mikehalmamoj**
- Put everything together
- Add integration tests
- Add additional validation
- Return OffenderBooking api model
- Handle SQLExceptions from Nomis
- Tidy up offender booking active flag

### CODE REVIEW

File: src/main/java/net/syscon/elite/web/handler/ControllerAdvice.java [hunk 1]
```diff
@@ -262,4 +267,50 @@
                         .developerMessage(e.getMessage())
                         .build());
     }
+
+    @ExceptionHandler(JpaSystemException.class)
```
Comment (mikehalmamoj) [inline src/main/java/net/syscon/elite/web/handler/ControllerAdvice.java:271]: This is currently handling the cell capacity error thrown from old Nomis, but that business rule will be duplicated in another PR.  However it seems like a good idea to catch any similar errors thrown out of Nomis and return a nice message - but each one we find should be duplicated in this app.

File: src/main/resources/db/migration/nomis/data/R__3_10__STAFF_USER_ACCOUNTS.sql [hunk 2]
```diff
@@ -19,3 +19,4 @@ INSERT INTO STAFF_USER_ACCOUNTS (USERNAME, STAFF_ID, STAFF_USER_TYPE, ID_SOURCE,
 INSERT INTO STAFF_USER_ACCOUNTS (USERNAME, STAFF_ID, STAFF_USER_TYPE, ID_SOURCE, WORKING_CASELOAD_ID) VALUES ('POM_USER', -21, 'GENERAL', 'USER', 'LEI');
 INSERT INTO STAFF_USER_ACCOUNTS (USERNAME, STAFF_ID, STAFF_USER_TYPE, ID_SOURCE, WORKING_CASELOAD_ID) VALUES ('PF_RO_USER', -22, 'GENERAL', 'USER', 'LEI');
 INSERT INTO STAFF_USER_ACCOUNTS (USERNAME, STAFF_ID, STAFF_USER_TYPE, ID_SOURCE, WORKING_CASELOAD_ID) VALUES ('UOF_COORDINATOR_USER', -23, 'GENERAL', 'USER', 'LEI');
+INSERT INTO STAFF_USER_ACCOUNTS (USERNAME, STAFF_ID, STAFF_USER_TYPE, ID_SOURCE, WORKING_CASELOAD_ID) VALUES ('WAI_USER', -25, 'GENERAL', 'USER', 'WAI');
```
Comment (mikehalmamoj) [inline src/main/resources/db/migration/nomis/data/R__3_10__STAFF_USER_ACCOUNTS.sql:22]: Needed to test @VerifyBookingAccess.

File: src/test/java/net/syscon/elite/api/resource/impl/OffenderMovementResourceApiTest_moveToCell.java [hunk 3]
```diff
@@ -72,19 +76,6 @@ public void validRequest_missingOptionalDateTime_isOk() {
         assertThat(response.getStatusCode()).isEqualTo(OK);
     }
 
-    @Test
-    public void validRequest_missingOptionalDateTime_defaultsToToday() {
```
Comment (mikehalmamoj) [inline src/test/java/net/syscon/elite/api/resource/impl/OffenderMovementResourceApiTest_moveToCell.java:76]: Moved to MovementUpdateServiceTest

File: src/test/resources/application-test.properties [hunk 4]
```diff
@@ -18,3 +18,5 @@ data.compliance.db.enable.parallel.hints=false
 
 spring.jpa.show-sql=true
 spring.jpa.properties.hibernate.format_sql=true
+
+spring.main.allow-bean-definition-overriding=true
```
Comment (mikehalmamoj) [inline src/test/resources/application-test.properties:22]: Required to override the Clock bean in integration tests.

File: src/test/resources/net/syscon/elite/api/resource/impl/offender_sentences.json [hunk 5]
```diff
@@ -429,7 +429,7 @@
   },
   "dateOfBirth": "1945-01-10",
   "agencyLocationDesc": "LEEDS",
-  "internalLocationDesc": "H-1"
+  "internalLocationDesc": "H-1-1"
```
Comment (mikehalmamoj) [inline src/test/resources/net/syscon/elite/api/resource/impl/offender_sentences.json:432]: Moved to a cell - H-1 is a wing but we make assumptions in the tests that offenders live in a cell.

File: src/main/java/net/syscon/elite/repository/jpa/model/OffenderBooking.java [hunk 6]
```diff
@@ -60,7 +58,10 @@
     private Offender offender;
 
     @Column(name = "LIVING_UNIT_ID")
-    private Long livingUnitId;
+    private Long assignedLivingUnitId;
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/repository/jpa/model/OffenderBooking.java:61]: Firing the question as I don't know.  Is this not mappable to an Object/Entity?

File: src/main/java/net/syscon/elite/repository/jpa/model/OffenderBooking.java [hunk 7]
```diff
@@ -77,7 +78,7 @@ public void add(final OffenderCourtCase courtCase) {
     }
 
     public boolean isActive() {
-        return ACTIVE_BOOKING.equals(bookingSequence);
+        return activeFlag != null && activeFlag.equals("Y");
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/repository/jpa/model/OffenderBooking.java:81]: +1

File: src/main/java/net/syscon/elite/service/BedAssignmentHistoryService.java [hunk 8]
```diff
@@ -15,11 +17,13 @@
...
+    public BedAssignmentHistoryService(final BedAssignmentHistoriesRepository repository) {
         this.repository = repository;
     }
 
-    public void add(Long bookingId, Long livingUnitId, String reasonCode, LocalDateTime time) {
+    @VerifyBookingAccess
+    @HasWriteScope
+    public void add(final Long bookingId, final Long livingUnitId, final String reasonCode, final LocalDateTime time) {
         final var maxSequence = repository.getMaxSeqForBookingId(bookingId);
         final var bookingAndSequence = new BedAssignmentHistory.BedAssignmentHistoryPK(bookingId, maxSequence + 1);
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/service/BedAssignmentHistoryService.java:28]: Is there no underlying (Oracle) sequence for this?  or is this how old NOMIS does it here?

File: src/main/java/net/syscon/elite/service/BookingService.java [hunk 9]
```diff
@@ -669,21 +673,32 @@ public MilitaryRecords getMilitaryRecords(final Long bookingId) {
     }
 
     @Transactional
+    @VerifyBookingAccess
+    @HasWriteScope
     public void updateLivingUnit(final Long bookingId, final Long livingUnitId) {
         var offenderBooking = offenderBookingRepository.findById(bookingId)
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/service/BookingService.java:679]: `final` ... meh.

File: src/main/java/net/syscon/elite/service/BookingService.java [hunk 10]
```diff
@@ -669,21 +673,32 @@ public MilitaryRecords getMilitaryRecords(final Long bookingId) {
...
                 offenderBooking.getLocation().getId().equals(location.getAgencyId()),
                 "Move to living unit in prison %s invalid for offender %s in prison %s",
                 location.getAgencyId(), offenderBooking.getOffender().getNomsId(), offenderBooking.getLocation().getId()
         );
-
-        offenderBooking.setLivingUnitId(livingUnitId);
-        offenderBookingRepository.save(offenderBooking);
-        log.info("Updated offender {} booking id {} to living unit id {}", offenderBooking.getOffender().getNomsId(), offenderBooking.getBookingId(), livingUnitId);
+        checkArgument(
+                location.getLocationType().equals("CELL"),
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/service/BookingService.java:698]: I'd consider pushing this sort of logic in the location / agency object perhaps.

File: src/main/java/net/syscon/elite/api/resource/impl/OffenderMovementsResourceImpl.java [hunk 11]
```diff
@@ -48,40 +44,13 @@ public CourtHearings getCourtHearings(final Long bookingId, final LocalDate from
     }
 
     @Override
-    public OffenderSummary moveToCell(final Long bookingId, final Long livingUnitId, final String reasonCode, final LocalDateTime dateTime) {
+    public OffenderBooking moveToCell(final Long bookingId, final Long livingUnitId, final String reasonCode, final LocalDateTime dateTime) {
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/api/resource/impl/OffenderMovementsResourceImpl.java:47]: I think this needs the `@ProxyUser` adding...

File: src/main/java/net/syscon/elite/service/BedAssignmentHistoryService.java (hunk 8 above)
Comment (Mjwillis) [reply src/main/java/net/syscon/elite/service/BedAssignmentHistoryService.java:28]: Yes second that - check for sequence

File: src/main/java/net/syscon/elite/service/MovementUpdateService.java [hunk 12]
```diff
@@ -16,34 +22,56 @@
...
-            return offenderSummary;
+        if (offenderBooking.getAssignedLivingUnitId().equals(livingUnitId)) {
+            return offenderBooking;
         }
 
-// TODO DT-235 Uncomment the updates - this is currently still a work in progress and we don't want to actually update anything yet
-//        bookingService.updateLivingUnit(bookingId, livingUnitId);
-//        bedAssignmentHistoryService.add(bookingId, livingUnitId, reasonCode, dateTime);
-        return getOffenderSummary(bookingId);
+        bookingService.updateLivingUnit(bookingId, livingUnitId);
```
Comment (Mjwillis) [inline src/main/java/net/syscon/elite/service/MovementUpdateService.java:54]: BTW - if you didn't do this update of booking does the trigger update anyway?

File: src/main/java/net/syscon/elite/service/BedAssignmentHistoryService.java (hunk 8 above)
Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/service/BedAssignmentHistoryService.java:28]: Old Nomis grabs the latest sequence and adds one - so I duplicated that code here.

File: src/main/java/net/syscon/elite/service/MovementUpdateService.java (hunk 12 above)
Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/service/MovementUpdateService.java:54]: No. All the trigger does is update the end dates on the previous bed assignment history record.  I suspect all of the other procedures are called from the Oracle forms...

Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/service/MovementUpdateService.java:54]: In fact this method performs the same updates as the process_release_bed procedure - update living unit and add to bed assignment histories.

File: src/main/java/net/syscon/elite/service/BedAssignmentHistoryService.java (hunk 8 above)
Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/service/BedAssignmentHistoryService.java:28]: Sorry that code mentioned above is found in procedure `ins_off_bed_assg_hty`
